lapack/testlapack/dorgr2.go

```diff
@@ -35,60 +37,72 @@ func Dorgr2Test(t *testing.T, impl Dorgr2er) {
 }
 
 func dorgr2Test(t *testing.T, impl Dorgr2er, rnd *rand.Rand, m, n, k, lda int) {
-	const tol = 1e-12
-	name := fmt.Sprintf("m=%v,n=%v,k=%v,lda=%v", m, n, k, lda)
+	const tol = 1e-14
 
-	if lda == 0 {
-		lda = n
-	}
+	name := fmt.Sprintf("m=%v,n=%v,k=%v,lda=%v", m, n, k, lda)
 
+	// Generate a random m×n matrix A.
 	a := randomGeneral(m, n, lda, rnd)
-	aCopy := cloneGeneral(a)
+
 	// Compute the RQ decomposition of A.
+	rq := cloneGeneral(a)
 	tau := make([]float64, m)
 	work := make([]float64, 1)
-	impl.Dgerqf(m, n, a.Data, a.Stride, tau, work, -1)
+	impl.Dgerqf(m, n, rq.Data, rq.Stride, tau, work, -1)
 	work = make([]float64, int(work[0]))
-	impl.Dgerqf(m, n, a.Data, a.Stride, tau, work, len(work))
+	impl.Dgerqf(m, n, rq.Data, rq.Stride, tau, work, len(work))
+
+	tauCopy := make([]float64, len(tau))
+	copy(tauCopy, tau)
 
-	// Generate the upper triangular matrix R from subarray A[m-k:m, n-m:n]
-	r := zeros(k, m, m)
-	for i := 0; i < k; i++ {
-		for j := 0; j < k; j++ {
-			ia := i + a.Rows - k
-			ja := j + a.Cols - k
-			jr := j + r.Cols - k
-			if i <= j {
-				r.Data[i*r.Stride+jr] = a.Data[ia*a.Stride+ja]
-			}
-		}
-	}
 	// Compute the matrix Q using Dorg2r.
-	impl.Dorgr2(m, n, k, a.Data, a.Stride, tau[m-k:m], work)
+	q := cloneGeneral(rq)
+	impl.Dorgr2(m, n, k, q.Data, q.Stride, tau[m-k:m], work)
+
 	if m == 0 {
 		return
 	}
```

Comment:
Does this happen this late for panic checks?

Reply:
Yes, to check for panics but also for modifications if there are non-zero-sized const input slices.

Reply:
> but also for modifications if there are non-zero-sized const input slices.

I'm confused about this. There are no test checks prior to this return, so if m is zero, the only thing that can be observed as a failure is a panic. What am I missing?

Reply:
Ah, sorry for not being clear, I meant this comment in general. You're right that here I postpone the quick return only for a panic check. In other routines there might also be returned errors or inputs that I check for modifications between the call and the quick return (for example, Dtbtrs).

Reply:
OK. Thanks.